tasks/EchoesToFrameConverterTask.cpp

```diff
@@ -0,0 +1,159 @@
+    return true;
+}
+bool EchoesToFrameConverterTask::startHook()
+{
+    if (!EchoesToFrameConverterTaskBase::startHook())
+        return false;
+    return true;
+}
+void EchoesToFrameConverterTask::updateHook()
+{
+    EchoesToFrameConverterTaskBase::updateHook();
+
+    base::Angle yaw_correction = base::Angle::fromRad(0);
+    base::samples::RigidBodyState sensor2ref_pose;
+    if (_sensor2ref_pose.read(sensor2ref_pose) != RTT::NoData) {
+        yaw_correction = base::Angle::fromRad(sensor2ref_pose.getYaw());
+    }
+
+    Radar radar_echo;
+    while (_echo.read(radar_echo, false) == RTT::NewData) {
+        if (m_current_sweep_size != radar_echo.sweep_length ||
+            m_current_num_angles != (2 * M_PI / radar_echo.step_angle.getRad()) ||
+            m_current_range != radar_echo.range) {
+            m_current_sweep_size = radar_echo.sweep_length;
+            m_current_num_angles = 2 * M_PI / radar_echo.step_angle.getRad();
+            m_current_range = radar_echo.range;
+            updateLookUpTable();
+        }
+
+        addEchoesToFrame(radar_echo, yaw_correction);
+    }
+    if (base::Time::now() - m_last_sample > _export_config.get().time_between_frames) {
+        publishFrame();
+    }
+}
+void EchoesToFrameConverterTask::errorHook()
+{
+    EchoesToFrameConverterTaskBase::errorHook();
+}
+void EchoesToFrameConverterTask::stopHook()
+{
+    EchoesToFrameConverterTaskBase::stopHook();
+}
+void EchoesToFrameConverterTask::cleanupHook()
+{
+    EchoesToFrameConverterTaskBase::cleanupHook();
+}
+
+void EchoesToFrameConverterTask::updateLookUpTable()
+{
+    LOG_INFO_S << "Updating LookUpTable";
+    auto config = _export_config.get();
+    m_lut.reset(new EchoToImageLUT(m_current_num_angles,
+        m_current_sweep_size,
+        config.beam_width,
+        config.output_image_size));
+    m_echoes = std::vector<uint8_t>(m_current_sweep_size * m_current_num_angles, 0);
+}
+
+void EchoesToFrameConverterTask::addEchoesToFrame(Radar const& echo,
+    base::Angle yaw_correction)
+{
+    LOG_INFO_S << "Adding echoes...";
+    double angle =
+        base::Angle::normalizeRad(echo.start_heading.getRad() + yaw_correction.getRad());
+    if (angle < 0) {
+        angle = 2 * M_PI + angle;
+    }
+
+    int start_angle_unit = round(angle / echo.step_angle.getRad());
+    int angles_in_a_frame = round(2 * M_PI / echo.step_angle.getRad());
+    LOG_DEBUG_S << "ANGLE MESSAGE-> start: " << start_angle_unit
+                << "YAW Correction: " << yaw_correction.getRad()
+                << " end: " << start_angle_unit + echo.sweep_timestamps.size()
+                << " Sweep length: " << echo.sweep_length;
+    for (int current_angle = 0;
+         current_angle < static_cast<int>(echo.sweep_timestamps.size());
+         current_angle++) {
+        LOG_DEBUG_S << "ANGLE LOOP-> Index: " << current_angle << " angle message: "
+                    << start_angle_unit + current_angle % angles_in_a_frame << "/"
```

Comment:
Especially since you end up having buggy debug output. This is a copy of the code below, meaning you *have* to make sure the two stay synchronized.

Fair enough to develop & debug. But I really don't think it's a good think to keep forever.

Reply:
Sorry ... I missed that one.